### DISCUSSION TITLE
help: CustomEvent possible?

### DISCUSSION THREAD

**Discussion by jnsw** [eclipse-milo/milo#911] [General]
I want to create a BaseEventTypeNode with Properties in BaseEventType to instance them later as an Event. But if I do so, it is not an abstract node like the other nodes already in BaseEventType and some "Bad Internal Error" occurs. If I am making UaObjectTypeNode, it can't cast UaObjectTypeNode to BaseEventTypeNode. What I am missing? Can someone provide some example code for creating an CustomEvent with properties, that I can use for example in startBogusEventNotifier() to see, if it works?

**Comment by kevinherron** [comment]
Can you post the full error and some code?

**Reply by jnsw** [reply]
The first error on creating BaseEventTypeNode to use it with event factory (creating inside of ExampleNamespace):
```
EventNsdto currentEventTypeNsdto = (EventNsdto) eventsMap.get(key); //just my DTO

NodeId currentEventType_NodeId = newNodeId(currentEventTypeNsdto.getNodeIdTypesNumber()); //its i=1027

BaseEventTypeNode currentEventTypeNode = new BaseEventTypeNode(
        getNodeContext(),
        currentEventType_NodeId,
        new QualifiedName(
                currentEventType_NodeId.getNamespaceIndex(),
                currentEventTypeNsdto.getBrowseName()),
        LocalizedText.english(currentEventTypeNsdto.getBrowseName()),
        LocalizedText.NULL_VALUE,
        uint(0),
        uint(0),
        true          //because it's abstract
);

//Adding reference to BaseEventType
currentEventTypeNode.addReference(new Reference(currentEventTypeNode.getNodeId(), Identifiers.HasSubtype, Identifiers.BaseEventType.expanded(), false));
getNodeManager().addNode(currentEventTypeNode);
```

Error comes when I change the "startBogusEventNotifier()" to use this new type instead of "Identifiers.BaseEventType"

```
// Post a bogus Event every couple seconds
            eventThread = new Thread(() -> {
                while (keepPostingEvents) {
                    try {
                        BaseEventTypeNode eventNode = getServer().getEventFactory().createEvent(
                                newNodeId(UUID.randomUUID()),
//                                Identifiers.BaseEventType
                                newNodeId(1027)
                        );

                        eventNode.setBrowseName(new QualifiedName(1, "okaycool"));
                        eventNode.setDisplayName(LocalizedText.english("fooooo"));
                        eventNode.setEventId(ByteString.of(new byte[]{0, 1, 2, 3}));
                        eventNode.setEventType(newNodeId(1027));
                        eventNode.setSourceNode(coffeemachine.getNodeId());                              //my source node
                        eventNode.setSourceName(coffeemachine.getDisplayName().getText());  //my source node
                        eventNode.setTime(DateTime.now());
                        eventNode.setReceiveTime(DateTime.NULL_VALUE);
                        eventNode.setMessage(LocalizedText.english("event messaaage!"));
                        eventNode.setSeverity(ushort(1));

                        //noinspection UnstableApiUsage
                        getServer().getEventBus().post(eventNode);

                        eventNode.delete();
                    } catch (Throwable e) {
                        GWLogger.error(String.format("Error creating EventNode: %s; %s", e.getMessage(), e));
                    }

                    try {
                        //noinspection BusyWait
                        Thread.sleep(2_000);
                    } catch (InterruptedException ignored) {
                        // ignored
                    }
                }
            }, "bogus-event-poster");
```

Error output on my logger: 

```
[33] err: Error creating EventNode: status=Bad_InternalError, description=An internal error occurred as a result of a programming or configuration error.; UaException: status=Bad_InternalError, message=An internal error occurred as a result of a programming or configuration error.
```
in debug mode i saw it crashes after returning the value of the nodeTree in createNode in NodeFactory, IntelliJ shows me, that there is a tree with the correct children, but just after that, it crashes with above error

![Bildschirmfoto 2022-01-26 um 22 18 36](https://user-images.githubusercontent.com/7985146/151248578-979b6df3-3eea-422d-a334-6c9c193f0942.png)


If I change `BaseEventTypeNode currentEventTypeNode` in the first code lines to `UaObjectTypeNode currentEventTypeNode` to have an abstract event node in my tree, it will crash by a cast error: 

```
[7] err: Error creating EventNode:class org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodecannot be cast toclass org.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNode(org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodeandorg.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNodeare in unnamed module of loader java.net.URLClassLoader @78047b92);java.lang.ClassCastException:class org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodecannot be cast toclass org.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNode(org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodeandorg.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNodeare in unnamed module of loader java.net.URLClassLoader @78047b92)
```

**Reply by jnsw** [reply]
recreated in clean 0.6.3 with following error - can you show me how I create costum EventTypes?
```
22:33:13.344 [bogus-event-poster] ERROR o.e.m.e.server.ExampleNamespace - Error creating EventNode: status=Bad_InternalError, description=An internal error occurred as a result of a programming or configuration error.
org.eclipse.milo.opcua.stack.core.UaException: status=Bad_InternalError, description=An internal error occurred as a result of a programming or configuration error.
	at org.eclipse.milo.opcua.sdk.server.nodes.factories.NodeFactory.createNodeTree(NodeFactory.java:193)
	at org.eclipse.milo.opcua.sdk.server.nodes.factories.NodeFactory.createNode(NodeFactory.java:98)
	at org.eclipse.milo.opcua.sdk.server.nodes.factories.EventFactory.createEvent(EventFactory.java:80)
	at org.eclipse.milo.examples.server.ExampleNamespace.lambda$startBogusEventNotifier$0(ExampleNamespace.java:261)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Reply by kevinherron** [reply]
You're missing a few steps it here. It sounds like you want to create your own custom event type that is a subtype of `BaseEventObjectType`.

You first need to create a `UaObjectTypeNode` that describes this event type (custom events are just custom objects, and custom objects need type definitions). Once you've created a type definition and registered it with the server's ObjectTypeManager then you can create an _instance_ of this event using the server's event factory.

What you're doing right now is creating an object instance and then telling the event factory to create another instance and incorrectly telling it the type definition of the instance you want to create is the NodeId of the object instance you just created.

**Reply by jnsw** [reply]
> You first need to create a UaObjectTypeNode that describes this event type (custom events are just custom objects, and custom objects need type definitions). Once you've created a type definition and registered it with the server's ObjectTypeManager then you can create an instance of this event using the server's event factory.

I did this in the second part of my first comment [here](https://github.com/eclipse/milo/discussions/911#discussioncomment-2054127), I created an UaObjectTypeNode like you mentioned now and it creates an casting error "**UaObjectNode cannot be cast to BaseEventTypeNode**"

`[7] err: Error creating EventNode:class org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodecannot be cast toclass org.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNode(org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodeandorg.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNodeare in unnamed module of loader java.net.URLClassLoader @78047b92);java.lang.ClassCastException:class org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodecannot be cast toclass org.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNode(org.eclipse.milo.opcua.sdk.server.nodes.UaObjectNodeandorg.eclipse.milo.opcua.sdk.server.model.nodes.objects.BaseEventTypeNodeare in unnamed module of loader java.net.URLClassLoader @78047b92)`

> incorrectly telling it the type definition of the instance you want to create is the NodeId of the object instance you just created.

damn, had another typo there, I just used newNodeId(1027) for fast using node ids, so it is the same nodeid but it still has this errors

**Reply by kevinherron** [reply]
`BaseEventTypeNode` is not a type definition. Look at its hierarchy, it's ultimately a UaObjectNode, not a UaObjectTypeNode.

**Reply by kevinherron** [reply]
Might need to register the object type and its constructor.

```
server.getObjectTypeManager().registerObjectType(...)
```

See https://github.com/eclipse/milo/blob/master/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/model/ObjectTypeInitializer.java for examples:
```
        objectTypeManager.registerObjectType(
            NodeId.parse("ns=0;i=11163")
                .reindex(namespaceTable, "http://opcfoundation.org/UA/"),
            BaseConditionClassTypeNode.class,
            BaseConditionClassTypeNode::new
        );
```

What this does is tell the server how to instantiate an instance of a given type definition. Without registering something it's trying to create raw UaObjectNode but you've cast the factory result to BaseEventTypeNode.

**Reply by jnsw** [reply]
Thanks, but now I am getting that error again.

> You first need to create a UaObjectTypeNode that describes this event type (custom events are just custom objects, and custom objects need type definitions). Once you've created a type definition and registered it with the server's ObjectTypeManager then you can create an instance of this event using the server's event factory.

You say I need `to create an UaObjectTypeNode` as custom event type and I already did that (first comment at the bottom), but having that UaObjectTypeNode generated as Type there gives me a **cast error**, if I want to `create an instance of this event using the server's event factory.` This error: `"UaObjectNode cannot be cast to BaseEventTypeNode"`. I don't see the difference in what you are saying and in what am I doing. Mind if you add an example to the example server we can adapt?

or: What am I missing?

**Reply by jnsw** [reply]
I will try that, thanks for helping!

**Reply by kevinherron** [reply]
This is all relatively undocumented because it's meant to be used primarily with UANodeSet models and code generators that don't exist / aren't public yet.

Creating and registering all this stuff by hand is a little tedious.

**Reply by jnsw** [reply]
It is, true, but it is a quite final server. or do you know any other open source java library I could use? Java is here the way I have to go.

**Reply by kevinherron** [reply]
There's not any other good Java options.

The OPC Foundation has abandoned their [UA-Java](https://github.com/OPCFoundation/UA-Java-Legacy) code, which was only a stack, and none of the "SDK", so you'd have to roll the entire service/application layer yourself.

Commercially there is the Prosys SDK if you don't mind paying some thousands of dollars.

**Reply by jnsw** [reply]
yeah, if its too complicated we would drive with prosys, but it a lot money for just that information and communication feature

**Reply by jnsw** [reply]
added 
```
        getServer().getObjectTypeManager().registerObjectType(
                newNodeId(1027),
                BaseEventTypeNode.class,
                BaseEventTypeNode::new
        );
```
and it works like a charm! thank you so much!

**Comment by jnsw** [comment]
sorry to ask again, but how do I write values to the custom defined properties? do I need to create special classes that extend BaseEventTypeNode or can I just BaseEventTypeNode.setProperty with a properly QualifiedProperty<> - but what do I need for a QualifiedProperty?

**Reply by kevinherron** [reply]
You need all the information about the property (name, datatype, etc...).

You can also just get the property node itself and then call setValue on it.

**Reply by jnsw** [reply]
> You can also just get the property node itself and then call setValue on it.

I am getting further, but I don't quite understand, I am now getting the Property-Node of my created event by using getPropertyNode(browseName) and am I getting the correct Node (as I can see in debug mode) to write my value to, but after publishing the event with `server.getEventBus().post(event);` it gets pushed but the property is still missing in UaExpert.


This is my created EventType I put into your factory.

<img width="1223" alt="Bildschirmfoto 2022-01-31 um 11 51 21" src="https://user-images.githubusercontent.com/7985146/151780983-40e8f486-a0a5-48c6-9f9f-8fcf93ba22a7.png">

<img width="1223" alt="Bildschirmfoto 2022-01-31 um 11 51 17" src="https://user-images.githubusercontent.com/7985146/151780970-62b9513b-fd52-4b8c-bee5-aaae8819697b.png">

but it seems not to add the property "CleaningType" which is an enum, because the manually added column is empty
<img width="1616" alt="Bildschirmfoto 2022-01-31 um 11 54 34" src="https://user-images.githubusercontent.com/7985146/151781408-23a8d35a-bd76-4998-b581-af0fb31a0e67.png">

**Reply by jnsw** [reply]
This is my generating EventType, which works fine like you can see in the screenshots above
```
       //loading DTO 
        EventNsdto currentEventTypeNsdto = (EventNsdto) InfoModelDescription.getTree().get(shortName); //TODO: Change to IMSN
        NodeId currentEventType_NodeId = newNodeId(currentEventTypeNsdto.getNodeIdTypesNumber());

        //Creating EventTypeNode
        UaObjectTypeNode currentEventTypeNode = new UaObjectTypeNode(
                getNodeContext(),
                currentEventType_NodeId,
                new QualifiedName(
                        currentEventType_NodeId.getNamespaceIndex(),
                        currentEventTypeNsdto.getBrowseName()),
                LocalizedText.english(currentEventTypeNsdto.getBrowseName()),
                LocalizedText.NULL_VALUE,
                uint(0),
                uint(0),
                true
        );

//Code for adding property (Code seperately below)

        //Adding Event to TypeManager
        getServer().getObjectTypeManager().registerObjectType(
                currentNode.getNodeId(),
                BaseEventTypeNode.class,
                BaseEventTypeNode::new
        );

        currentNode.addReference(new Reference(currentNode.getNodeId(), Identifiers.HasSubtype, Identifiers.BaseEventType.expanded(), false));

        getNodeManager().addNode(currentNode);

```

Creating and adding Property: This is how I add an example property to that event node above
 ```
      //Creating Property for EventTypeNode
        PropertyNsdto currentPropertyNsdto = (PropertyNsdto) InfoModelDescription.COFFEEMACHINE_INFOMODEL_MAP.get(shortName);

        NodeId propertyNodeId = newNodeId(currentPropertyNsdto.getNodeIdTypesNumber());

        //Create PropertyNode
        PropertyTypeNode propertyNode = new PropertyTypeNode(
            getNodeContext(), 
            propertyNodeId, 
            new QualifiedName(propertyNodeId.getNamespaceIndex(), currentPropertyNsdto.getBrowseName()), 
            LocalizedText.english(currentPropertyNsdto.getBrowseName()), 
            LocalizedText.NULL_VALUE, 
            uint(0), 
            uint(0)
        );

        propertyNode.setDataType(
                currentPropertyNsdto.isCustomDataType() ? 
                        newNodeId(currentPropertyNsdto.getCustomDataTypeIdentifier()) : 
                currentPropertyNsdto.getDataType().expanded().toNodeId(getNodeContext().getNamespaceTable()).orElse(NodeId.NULL_VALUE));
        
        propertyNode.setValueRank(ValueRank.Scalar.getValue()); 
        propertyNode.setArrayDimensions(null);
        propertyNode.setMinimumSamplingInterval(currentPropertyNsdto.getMinimumSamplingInterval());
        propertyNode.setAccessLevel(AccessLevel.toValue(currentPropertyNsdto.getAccessLevel()));
        propertyNode.setUserAccessLevel(AccessLevel.toValue(currentPropertyNsdto.getAccessLevel()));
       
        //adding references that this node is a property
        propertyNode.addReference(new Reference(propertyNode.getNodeId(), Identifiers.HasTypeDefinition, Identifiers.PropertyType.expanded(), true));
        if (currentPropertyNsdto.getMandatory()) 
                propertyNode.addReference(new Reference(propertyNode.getNodeId(), Identifiers.HasModellingRule, Identifiers.ModellingRule_Mandatory.expanded(), true));

        //set initial value to zero
        propertyNode.setValue(new DataValue(new Variant(null)));
        //adding to tree
        getNodeContext().getNodeManager().addNode(propertyNode);

        //adding propertyreferences
        currentEventTypeNode.addReference(new Reference(currentEventTypeNode.getNodeId(), Identifiers.HasProperty, createdProperty.getNodeId().expanded(), true));
        createdProperty.addReference(new Reference(createdProperty.getNodeId(), Identifiers.HasProperty, currentEventTypeNode.getNodeId().expanded(), false));
```

**Reply by kevinherron** [reply]
After you instantiate an event of this type using the EventFactory are you setting a value for this property? It looks like you have a null value by default.

**Reply by jnsw** [reply]
> You can also just get the property node itself and then call setValue on it.

i did this: 
```
	@Override
	public void invokeCleaningEvent(int severity, String message, CleaningTypeEnumeration cleaningTypeEnumeration){
		try {
			NodeId cleaningEvent_NodeId = someNodeIdGetter;
			NodeId cleaningEvent_CleaningType_NodeId = someNodeIdGetter;

			BaseEventTypeNode event = createEvent(cleaningEvent_NodeId); //createEvent is below

			event.setSeverity(ushort(severity));

			event.setMessage(LocalizedText.english(message));

			QualifiedProperty<CleaningTypeEnumeration> cleaningTypeEnumerationQualifiedProperty = new QualifiedProperty(
				server.getServerNamespace().getNamespaceUri(), 
				"CleaningType", 
				cleaningEvent_CleaningType_NodeId.expanded(), 
				-1, 
				CleaningTypeEnumeration.class
			);

			event.setProperty(cleaningTypeEnumerationQualifiedProperty, cleaningTypeEnumeration);

			server.getEventBus().post(eventNode);

			eventNode.delete();
		} catch (NullPointerException e) {
			GWLogger.error("Got NullPointer in invokeCleaningEvent");
		}
	}
```



```	
	public BaseEventTypeNode createEvent(NodeId eventTypeId){
		BaseEventTypeNode event = null;
		try {
			UaNode serverNode = server
				.getAddressSpaceManager()
				.getManagedNode(Identifiers.Server)
				.orElse(null);

			event = server.getEventFactory().createEvent(
				new NodeId(server.getServerNamespace().getNamespaceIndex(), UUID.randomUUID()),
				eventTypeId
			);

			// event.setBrowseName(new QualifiedName(1, "TestNodeEvent"));
			// event.setDisplayName(LocalizedText.english("TestNodeEvent"));
			event.setEventId(ByteString.of(new byte[]{0, 1, 2, 3}));
			// event.setEventType(Identifiers.BaseEventType);
			event.setEventType(eventTypeId);
			event.setSourceNode(serverNode.getNodeId());
			event.setSourceName(serverNode.getDisplayName().getText());
			event.setTime(DateTime.now());
			event.setReceiveTime(DateTime.NULL_VALUE);
			event.setMessage(LocalizedText.NULL_VALUE);
			event.setSeverity(ushort(1));
		} catch (Throwable e) {
			GWLogger.error(String.format("Error creating EventNode: %s; %s", e.getMessage(), e));
		}
	}
```

**Comment by jnsw** [comment]
Solved: I used the wrong namespace_uri - fyi

_______

Okay, I now started to create Classes like
"ProductDispenseEventType" filled with QualifiedProperties and getXYNode, getXY, and setXY method-stubs
and
"ProductDispenseEventTypeNode" and filled with the methods.

I create the Nodes and add them to the type-tree and after that I am adding them to the ObjectTypeManager with
```
getServer().getObjectTypeManager().registerObjectType(
        currentNode.getNodeId(),
        ProductDispenseEventTypeNode.class,
        ProductDispenseEventTypeNode::new
);
```

But when I am invoking an event it is still just an SimpleEvent without the self defined properties. 

If I am invoking some other event with properties like "ConditionType" I am getting the specific properties - am I missing something?

**Comment by devaskim** [comment]
@jnsw @kevinherron Is there any success story (example) of adding custom event type to server?  (at least with single integer field)
